src/txrequest.cpp

```diff
@@ -0,0 +1,685 @@
+    const uint64_t m_sequence : 59;
+    /** Whether the request is preferred. */
+    const bool m_preferred : 1;
+    /** Whether this is a wtxid request. */
+    const bool m_is_wtxid : 1;
+
+    /** What state this announcement is in. This is a uint8_t instead of a State to silence a GCC warning. */
+    uint8_t m_state : 3;
+
+    /** Convert the m_state variable to a State enum. */
+    State GetState() const { return State(m_state); }
+    /** Convert a State to a uint8_t and store it in m_state. */
+    void SetState(State state) { m_state = uint8_t(state); }
+
+    /** Whether this entry is selected. There can be at most 1 selected peer per txhash. */
+    bool IsSelected() const
+    {
+        return GetState() == State::CANDIDATE_BEST || GetState() == State::REQUESTED;
+    }
+
+    /** Whether this entry is waiting for a certain time to pass. */
+    bool IsWaiting() const
+    {
+        return GetState() == State::REQUESTED || GetState() == State::CANDIDATE_DELAYED;
+    }
+
+    /** Whether this entry can feasibly be selected if the current IsSelected() one disappears. */
+    bool IsSelectable() const
+    {
+        return GetState() == State::CANDIDATE_READY || GetState() == State::CANDIDATE_BEST;
+    }
+
+    /** Construct a new entry from scratch, initially in CANDIDATE_DELAYED state. */
+    Entry(const GenTxid& gtxid, uint64_t peer, bool preferred, std::chrono::microseconds reqtime,
+        uint64_t sequence) :
+        m_txhash(gtxid.GetHash()), m_time(reqtime), m_peer(peer), m_sequence(sequence), m_preferred(preferred),
+        m_is_wtxid(gtxid.IsWtxid()), m_state(uint8_t(State::CANDIDATE_DELAYED)) {}
+};
+
+/** A functor with embedded salt that computes priority of an announcement.
+ *
+ * Lower priorities are selected first.
+ */
+class PriorityComputer {
+    const uint64_t m_k0, m_k1;
+public:
+    explicit PriorityComputer(bool deterministic) :
+        m_k0{deterministic ? 0 : GetRand(0xFFFFFFFFFFFFFFFF)},
+        m_k1{deterministic ? 0 : GetRand(0xFFFFFFFFFFFFFFFF)} {}
+
+    uint64_t operator()(const uint256& txhash, uint64_t peer, bool preferred) const
+    {
+        uint64_t low_bits = CSipHasher(m_k0, m_k1).Write(txhash.begin(), txhash.size()).Write(peer).Finalize() >> 1;
+        return low_bits | uint64_t{!preferred} << 63;
+    }
+
+    uint64_t operator()(const Entry& entry) const
+    {
+        return operator()(entry.m_txhash, entry.m_peer, entry.m_preferred);
+    }
+};
+
+// Definitions for the 3 indexes used in the main data structure.
+//
+// Each index has a By* type to identify it, a Entry* data type to represent the view of Entry it is sorted by,
+// and an Entry*Extractor type to convert an Entry into the Entry* view type.
+// See https://www.boost.org/doc/libs/1_54_0/libs/multi_index/doc/reference/key_extraction.html#key_extractors
+// for more information about the key extraction concept.
+
+// The ByPeer index is sorted by (peer, state == CANDIDATE_BEST, txhash)
```

Comment:
That would make `GetRequestable` be `O(nr_announcements)` instead of `O(nr_best)` which I think makes the complexity blow out in worst case scenarios.

Reply:
Yes, makes sense. Thanks!

Reply:
Yes, that's exactly the reason.